pkg/controllers/reconcilers.go

```diff
@@ -54,14 +54,34 @@ type Config struct {
 	Tracker   tracker.Tracker
 }
 
+// ObjectType represents the type of a Kubernetes object. Fields of type ObjectType are
+// initialized with an empty object and not modified subsequently.
+type ObjectType struct {
+	runtime.Object
+}
+
+// NewObjectType constructs an ObjectType from the given empty object.
+func NewObjectType(obj runtime.Object) ObjectType {
+	return ObjectType{obj}
+}
+
+// asObject returns a deep copy of the object type as an apis.Object.
+func (ot ObjectType) asObject() apis.Object {
+	return ot.DeepCopyObject().(apis.Object)
```

Comment:
by convention, methods do not return copies. It's the responsibility of the caller to create a copy before inflicting mutations that should be localized.